### ISSUE TITLE
optimize_read_in_order + subquery -- wrong query result.

### ISSUE THREAD

**Issue by den-crane** [ClickHouse/ClickHouse#7354]
19.16.1.1414

```
set max_thread = 36;
set compile_expressions = 1;
SET optimize_read_in_order = 1;

create TABLE attr_listt1 (l String) Engine= Memory;
insert into attr_listt1 values ('1');
CREATE TABLE TESTTABLE4 (_id UInt64 , _date Date ,
  pt String, ts DateTime, attr_list String 
) ENGINE = MergeTree() PARTITION BY (pt, toStartOfWeek(_date))
ORDER BY (_id, ts);

insert into TESTTABLE4 values 
  (466000019,'2019-11-08','1','2019-10-16 14:40:45','1'),
  (466000018,'2019-11-08','0','2019-10-16 14:40:44','1');

SELECT _id
FROM TESTTABLE4
PREWHERE attr_list IN (select l from attr_listt1)
ORDER BY _id DESC
LIMIT 10

┌───────_id─┐
│ 466000018 │
└───────────┘
┌───────_id─┐
│ 466000019 │
└───────────┘

second execution
┌───────_id─┐
│ 466000019 │
└───────────┘
┌───────_id─┐
│ 466000018 │
└───────────┘


```

Whole story https://groups.google.com/d/msg/clickhouse/bUEoZC4S6-A/xiIcxJ_OCAAJ

<img width="963" alt="Screen Shot 2019-10-16 at 12 28 14 PM" src="https://user-images.githubusercontent.com/19737682/66934210-87343780-f010-11e9-98b5-2bfc30228f5f.png">

**Comment by den-crane** [comment]
Simplified. 
It seams just need to have data in different partitions
```

set compile_expressions = 0;
set max_threads = 2;

drop table TESTTABLE4
CREATE TABLE TESTTABLE4 (_id UInt64 , pt String, l String 
) ENGINE = MergeTree() PARTITION BY (pt) ORDER BY (_id);
insert into TESTTABLE4 values (0,'1','1'), (1,'0','1');
drop table attr_listt1;
create TABLE attr_listt1 (l String) Engine= Memory;
insert into attr_listt1 values ('1');

SELECT _id FROM TESTTABLE4 PREWHERE l IN (select l from attr_listt1)
ORDER BY _id DESC LIMIT 10

┌─_id─┐
│   0 │
└─────┘
┌─_id─┐
│   1 │
└─────┘

2 rows in set. Elapsed: 0.002 sec.


┌─_id─┐
│   1 │
└─────┘
┌─_id─┐
│   0 │
└─────┘

2 rows in set. Elapsed: 0.002 sec.
```

**Comment by CurtizJ** [comment]
FIxed in master #7371.
